File: q208.js
class TrieNode {
  constructor() {
    this.children = {}
    this.endOfWord = false
  }
}

class Trie {
  constructor() {
    this.root = new TrieNode()
  }

  insert(word) {
    let cur = this.root

    for (let c of word) {
      if (!cur.children.hasOwnProperty(c)) {
        cur.children[c] = new TrieNode()
      }

      cur = cur.children[c]
    }

    cur.endOfWord = true
  }

  search(word) {
    let cur = this.root

    for (let c of word) {
      if (!cur.children.hasOwnProperty(c)) {
        return false
      }

      cur = cur.children[c]
    }

    return cur.endOfWord
  }

  startWith(word) {
    let cur = this.root

    for (let c of word) {
      if (!cur.children.hasOwnProperty(c)) {
        return false
      }

      cur = cur.children[c]
    }

    return true
  }
}

let trie = new Trie()
trie.insert("apple")
trie.insert("pine")
console.log(trie.search("apple"))
console.log(trie.startWith("pine"))

File: q208.py
class TrieNode:
  def __init__(self):
    self.children = {}
    self.endOfWord = False

class Trie:
  def __init__(self):
    self.root = TrieNode()

  def insert(self, word: str) -> None:
    cur = self.root

    for c in word:
      if c not in cur.children:
        cur.children[c] = TrieNode()
      cur = cur.children[c]
    cur.endOfWord = True


  def search(self, word: str) -> bool:
    cur = self.root

    for c in word:
      if c not in cur.children:
        return False
      cur = cur.children[c]

    return cur.endOfWord

  def startWith(self, word: str) -> bool:
    cur = self.root
    
    for c in word:
      if c not in cur.children:
        return False
      cur = cur.children[c]
    
    return True

trie = Trie()
trie.insert("apple")
print(trie.search("apple"))
print(trie.startWith("ap"))